dash_auth/oidc_auth.py

```diff
@@ -0,0 +1,298 @@
+from authlib.integrations.flask_client import OAuth
+from dash_auth.auth import Auth
+from flask import redirect, request, session, url_for
+from werkzeug.routing import Map, Rule
+
+if TYPE_CHECKING:
+    from authlib.integrations.flask_client.apps import (
+        FlaskOAuth1App, FlaskOAuth2App
+    )
+
+
+class OIDCAuth(Auth):
+    """Implements auth via OpenID."""
+
+    def __init__(
+        self,
+        app: dash.Dash,
+        secret_key: str = Optional[None],
+        force_https_callback: Optional[Union[bool, str]] = None,
+        login_route: str = "/oidc/<idp>/login",
+        logout_route: str = "/oidc/logout",
+        callback_route: str = "/oidc/<idp>/callback",
+        idp_selection_route: str = None,
+        log_signins: bool = False,
+        public_routes: Optional[list] = None,
+        logout_page: str = None,
+    ):
+        """Secure a Dash app through OpenID Connect.
+
+        Parameters
+        ----------
+        app : Dash
+            The Dash app to secure
+        secret_key : str, optional
+            A string to protect the Flask session, by default None.
+            Generate a secret key in your Python session
+            with the following commands:
+            >>> import os
+            >>> import base64
+            >>> base64.b64encode(os.urandom(30)).decode('utf-8')
+            Note that you should not do this dynamically:
+            you should create a key and then assign the value of
+            that key in your code.
+        force_https_callback : Union[bool, str], optional
+            Whether to force redirection to https, by default None
+            This is useful when the HTTPS termination is upstream of the server
+            If a string is passed, this will check for the existence of
+            an envvar with that name and force https callback if it exists.
+        client_kwargs : dict, optional
+            Keyword arguments passed to the OAuth client
+        login_route : str, optional
+            The route for the login function, it requires a <idp>
+            placeholder, by default "/oidc/<idp>/login".
+        logout_route : str, optional
+            The route for the logout function, by default "/oidc/logout".
+        callback_route : str, optional
+            The route for the OIDC redirect URI, it requires a <idp>
+            placeholder, by default "/oidc/<idp>/callback".
+        idp_selection_route : str, optional
+            The route for the IDP selection function, by default None
+        log_signins : bool, optional
+            Whether to log signins, by default False
+        **kwargs
+            Additional keyword arguments are passed to oauth.register.
+
+        Raises
+        ------
+        Exception
+            Raise an exception if the app.server.secret_key is not defined
+        """
+        super().__init__(app, public_routes=public_routes)
+        if force_https_callback is not None:
```

Comment:
The branching would benefit from being flattened. I think the first if branch should be` if isinstance(force_https_callback, str)`